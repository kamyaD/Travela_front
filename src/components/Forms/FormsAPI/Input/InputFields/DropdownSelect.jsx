import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import selectDropdownIcon from '../../../../../images/icons/form_select_dropdown.svg';
import DropdownOptions from './DropdownOptions';

export default class DropdownSelect extends Component {
  static propTypes = {
    className: PropTypes.string,
    error: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    choices: PropTypes.arrayOf(PropTypes.string).isRequired,
    value: PropTypes.string.isRequired,
    size: PropTypes.string.isRequired
  };

  static defaultProps = {
    className: '',
    error: ''
  };

  state = {
    dropdownOpen: false
  };

  getSelectOptions(choices) {
    return (
      <DropdownOptions
        items={choices}
        handleClick={this.handleOptClick}
        getDropdownStatus={this.getDropdownStatus}
      />
    );
  }

  getDropdownStatus = () => {
    const { dropdownOpen } = this.state;
    const dropdownStatus = dropdownOpen ? 'active' : 'inactive';
    return dropdownStatus;
  };

  getPropsObject = () => {
    const { className, error } = this.props;
    const _props = { ...this.props };
    ['labelNote', 'selectOptions', 'handleSelect', 'selectedDate'].map(
      item => delete _props[item]
    );

    const props = {
      ..._props,
      onClick: this.handleToggleDropdown,
      onKeyPress: () => {},
      role: 'button',
      tabIndex: '0',
      className: `input select ${error ? 'error' : ''} ${
        className ? className : ''
      }`
    };
    return props;
  };

  handleOptClick = choice => {
    const { onChange, handleDropDown } = this.props;
    onChange(choice);
    this.handleToggleDropdown();
    handleDropDown ? handleDropDown(this.props, choice) : null;
  };

  handleToggleDropdown = e => {
    const { dropdownOpen } = this.state;
    const { onClick } = this.props;
    onClick && onClick(e);
    this.setState(() => ({
      dropdownOpen: !dropdownOpen
    }));
  };

  valueToChoiceLabel = (choices, value) => {
    let choice =
      choices.filter(choice => {
        return (choice.value || choice) === value;
      })[0] || '';
    choice = !choice ? ' ' : choice;
    return choice.label || choice || value;
  };

  render() {
    const { choices, value, size } = this.props;
    const options = this.getSelectOptions(choices);
    return (
      <div style={{ position: 'relative' }}>
        <div {...this.getPropsObject()}>
          <div className="value" style={{ width: size }}>
            {this.valueToChoiceLabel(choices, value)}
            <div className={`select-dropdown ${this.getDropdownStatus()}`}>
              <img src={selectDropdownIcon} alt="icn" />
            </div>
          </div>
          {options}
        </div>
      </div>
    );
  }
}
