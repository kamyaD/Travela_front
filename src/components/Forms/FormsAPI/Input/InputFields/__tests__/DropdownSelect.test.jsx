import React from 'react';
import { shallow } from 'enzyme';
import DropdownSelect from '../DropdownSelect';

describe('<DropdownSelect />', () => {
  const choices = ['manager', 'associate'];
  const onChange = () => {};
  it('renders fine', () => {
    const wrapper = shallow(<DropdownSelect value="" size="245px" choices={choices} onChange={onChange} />);
    expect (wrapper).toMatchSnapshot();
  });
  it('test onClick event of choices in dropdown', () => {
    const wrapper = mount(<DropdownSelect value="" size="245px" choices={choices} onChange={onChange} />);
    expect (wrapper.find('div.input')).toHaveLength(1);
    wrapper.find('div.input').simulate('click');
    wrapper.find('#choice').at(0).simulate('click');
    expect (wrapper.state('dropdownOpen')).toBe(false);
  });
  it('clone should call handleOptClick when clicked', () => {      
    const wrapper = mount(<DropdownSelect value="" size="245px" choices={choices} onChange={onChange} />);
    const spy = jest.spyOn(wrapper.instance(), 'handleOptClick');
    wrapper.update();
    wrapper.find('div.input').simulate('click');
    wrapper.find('div#choice').at(0).simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

});
