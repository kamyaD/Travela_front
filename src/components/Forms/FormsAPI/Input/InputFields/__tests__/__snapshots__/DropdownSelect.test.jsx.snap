// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownSelect /> renders fine 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    choices={
      Array [
        "manager",
        "associate",
      ]
    }
    className="input select  "
    error=""
    onChange={[Function]}
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="value"
      style={
        Object {
          "width": undefined,
        }
      }
    >
       
      <div
        className="select-dropdown inactive"
      >
        <img
          alt="icn"
          src="form_select_dropdown.svg"
        />
      </div>
    </div>
    <DropdownOptions
      getDropdownStatus={[Function]}
      handleClick={[Function]}
      items={
        Array [
          "manager",
          "associate",
        ]
      }
    />
  </div>
</div>
`;
